src/NServiceBus.AcceptanceTests/Sagas/When_saga_handles_messages_concurrently.cs

```diff
@@ -0,0 +1,154 @@
+            public bool SagaCompleted { get; set; }
+        }
+
+        public class ConcurrentHandlerEndpoint : EndpointConfigurationBuilder
+        {
+            public ConcurrentHandlerEndpoint()
+            {
+                EndpointSetup<DefaultServer>(b =>
+                {
+                    b.LimitMessageProcessingConcurrencyTo(2);
+                    b.Recoverability().Immediate(immediate => immediate.NumberOfRetries(3));
+                });
+            }
+
+            public class ConcurrentlyStartedSaga : Saga<ConcurrentlyStartedSagaData>,
+                IAmStartedByMessages<StartMessageTwo>,
+                IAmStartedByMessages<StartMessageOne>
+            {
+                public Context Context { get; set; }
+
+                public async Task Handle(StartMessageOne message, IMessageHandlerContext context)
+                {
+                    Data.Placed = true;
+                    await context.SendLocal(new SuccessfulProcessing
+                    {
+                        SagaId = Data.Id,
+                        Type = nameof(StartMessageOne)
+                    });
+                    await CheckForCompletion(context);
+                }
+
+                public async Task Handle(StartMessageTwo message, IMessageHandlerContext context)
+                {
+                    Data.Billed = true;
+                    await context.SendLocal(new SuccessfulProcessing
+                    {
+                        SagaId = Data.Id,
+                        Type = nameof(StartMessageTwo)
+                    });
+                    await CheckForCompletion(context);
+                }
+
+                protected override void ConfigureHowToFindSaga(SagaPropertyMapper<ConcurrentlyStartedSagaData> mapper)
+                {
+                    mapper.ConfigureMapping<StartMessageOne>(msg => msg.SomeId).ToSaga(saga => saga.OrderId);
+                    mapper.ConfigureMapping<StartMessageTwo>(msg => msg.SomeId).ToSaga(saga => saga.OrderId);
+                }
+
+                Task CheckForCompletion(IMessageHandlerContext context)
+                {
+                    if (Data.Billed && Data.Placed)
+                    {
+                        MarkAsComplete();
+                        Context.SagaCompleted = true;
+                    }
+                    return Task.FromResult(0);
+                }
+            }
+
+            public class ConcurrentlyStartedSagaData : ContainSagaData
+            {
+                public virtual string OrderId { get; set; }
+                public virtual bool Placed { get; set; }
+                public virtual bool Billed { get; set; }
+            }
+
+            public class LogSuccessfulHandler : IHandleMessages<SuccessfulProcessing>
+            {
+                public Context Context { get; set; }
+
+                public Task Handle(SuccessfulProcessing message, IMessageHandlerContext context)
```

Comment:
If I remember correctly, this is used to ensure we're not corrupting the test-context from messages which are rolled back, right? Should we add a comment to make that more clear for future readers?